examples/generative/ddpm.py

```diff
@@ -0,0 +1,682 @@
+        )
+        posterior_variance = self._extract(self.posterior_variance, t, x_t_shape)
+        posterior_log_variance_clipped = self._extract(
+            self.posterior_log_variance_clipped, t, x_t_shape
+        )
+        return posterior_mean, posterior_variance, posterior_log_variance_clipped
+
+    def p_mean_variance(self, pred_noise, x, t, clip_denoised=True):
+        x_recon = self.predict_start_from_noise(x, t=t, noise=pred_noise)
+        if clip_denoised:
+            x_recon = tf.clip_by_value(x_recon, self.clip_min, self.clip_max)
+        
+        model_mean, posterior_variance, posterior_log_variance = self.q_posterior(
+            x_start=x_recon, x_t=x, t=t
+        )
+        return model_mean, posterior_variance, posterior_log_variance
+
+    def p_sample(self, pred_noise, x, t, clip_denoised=True):
+        """Sample from the diffuison model.
+
+        Args:
+            pred_noise: Noise predicted by the diffusion model
+            x: Samples at a given timestep for which the noise was predicted
+            t: Current timestep
+            clip_denoised (bool): Whether to clip the predicted noise
+                within the specified range or not.
+        """
+        model_mean, _, model_log_variance = self.p_mean_variance(
+            pred_noise, x=x, t=t, clip_denoised=clip_denoised
+        )
+        noise = tf.random.normal(shape=x.shape, dtype=x.dtype)
+        # No noise when t == 0
+        nonzero_mask = tf.reshape(
+            1 - tf.cast(tf.equal(t, 0), tf.float32), [tf.shape(x)[0], 1, 1, 1]
+        )
+        return model_mean + nonzero_mask * tf.exp(0.5 * model_log_variance) * noise
+
+
+"""
+## Network Architecture
+
+U-Net, originally developed for semantic segmentation, is the architecture that is
+widely used for implementing diffusion models but with some slight modifications:
+
+1. The network accepts two inputs: Image and time step
+2. Self-attention between the convolution blocks once we reach a specific resolution
+(16x16 in the paper)
+3. Group Normalization instead of weight normalization
+
+We will implement most of the things as used in the original paper. We will use `swish`
+activation function throughout the network. We will use variance scaling kernel
+initializer. 
+
+The only difference here is the number of groups used for GroupNormalization layer.
+For the flowers dataset, I found that a value of `groups=8` produces better results
+compared to the default value of `groups=32`. Dropout is optional and should be
+used where chances of over fitting is high. In the paper, the authors used dropout
+only when training on CIFAR10.
+"""
+
+
+# Activation function to use
+act_fn = keras.activations.swish
+
+
+# Kernel initializer to use
+def kernel_init(scale):
+    if scale == 0.0:
+        scale = 1e-10
+    return keras.initializers.VarianceScaling(
+        scale, mode="fan_avg", distribution="uniform"
+    )
+
+
+def AttentionBlock(units, groups=8):
```

Comment:
Given that you use a lot of raw TF ops rather than just Keras layers, I recommend making this a `Layer` subclass.